keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,727 @@
++ _BucketClass (BC)_ - a cluster-scoped custom resource containing fields defining the provisioner and an immutable parameter set for creating new buckets.
++ _BucketRequest (BR)_ - a user-namespaced custom resource representing a request for a new backend bucket, or access to an existing bucket.
++ _COSI_ - Container _Object_ Store Interface, modeled after CSI.
++ _cosi-node-adapter_ - a pod per node which receives Kubelet gRPC _NodeStageVolume_ and _NodeUnstageVolume_ requests, ensures the target bucket has been provisioned, and notifies the kubelet when the pod can be run.
++ _driver_ - a container (usually paired with a sidecar container) that is responsible for communicating with the underlying object store to create, delete, grant access to, and revoke access from buckets. Drivers talk gRPC and need no knowledge of Kubernetes. Drivers are typically written by storage vendors, and should not be given any access outside of their namespace.
++ _driverless_ - a use-case where no driver exists to support the backend object store. Some COSI automation may still be provided, but backend operations such as creating a bucket or minting credentials will not occur.
++ _greenfield bucket_ - a new backend bucket created by COSI.
++ _green-to-brownfield_ - a use-case where COSI creates a new bucket on behalf of a user, and then supports ways for other users in the cluster to share this bucket.
++ _provisioner_ - a generic term meant to describe the combination of a sidecar and driver. "Provisioning" a bucket can mean creating a new bucket or granting access to an existing bucket.
++ _sidecar_ - a Kubernetes-aware container (usually paired with a driver) that is responsible for fullfilling COSI requests with the driver via gRPC calls. The sidecar's access can be restricted to the namespace where it resides, which is expected to be the same namespace as the provisioner. The COSI sidecar is provided by the Kubernetes community.
++ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.
+
+# Proposal
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.
+
+There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    - cosi.io/provisioner: [1]
+  finalizers:
+    - cosi.io/finalizer [2]
+spec:
+  protocol:
+    name: [3]
+    version: [4]
```

Comment:
I agree with this sentiment in general. However, this field is filled by COSI in some cases and we think it is better to have COSI fill an unfilled field rather than update an existing one.

Reply:
To simplify the prototype and avoid the controversy of what binding logic will look like for now, I'm fine with removing version from prototype from all objects. For alpha/beta, we can discuss and reintroduce.